haskell-style.md

```diff
@@ -391,3 +412,134 @@ import qualified Data.List.NonEmpty as NE
 
 import qualified Data.Sequence as Seq
 ```
+
+### Import groups
+
+Imports should be divided into groups:
+
+1. Prelude (when explicit)
+1. Un-`qualified` dependency libraries
+1. `qualified` dependency libraries
+1. Un-`qualified` FrontRow libraries
+1. `qualified` FrontRow libraries
+1. Un-`qualified` modules from this library
+1. `qualified` modules from this libraries
+
+Aside from the empty line between groups, there should be no other whitespace.
+
+```haskell
+-- Bad
+-- What's coming from where?
+import ClassyPrelude
+import Network.AWS.S3
+import Network.AWS
+import Data.Text (Text)
+import Json
+import qualified Data.Text as T
+import Control.Lens
+import Unit
+import Data.Conduit
+import TextAssets.S3
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+-- Bad
+-- You can't quickly highlight and :sort things
+import ClassyPrelude
+
+import Network.AWS.S3
+import Network.AWS
+import Data.Text (Text)
+import qualified Data.Text as T
+import Control.Lens
+import Data.Conduit
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+import TextAssets.S3
+
+import Json (Json)
+import qualified Json as J
+import Unit
+
+-- Good
+import ClassyPrelude
+
+import Control.Lens
+import Data.Conduit
+import Data.Map (Map)
+import Data.Set (Set)
+import Data.Text (Text)
+import Network.AWS
+import Network.AWS.S3
+
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Data.Text as T
+
+import TextAssets.S3
+
+import Json (Json)
+import Unit
+
+import qualified Json as J
+```
+
+## Exports
+
+Use sorted, multi-line exports. There are two exceptions to this rule:
+
+1. A single-line export for `Main`, when it only exports `main`:
+
+   ```haskell
+   module Main (main) where
+   ```
+
+1. If the order of exports matters in your desired Haddock output, you may
+   violate sorting to achieve it.
+
+Otherwise:
+
+```haskell
+-- Bad
+module Driver (scienceOptions, socialStudiesOptions, mainWith) where
+
+-- Good
+module Driver
+  ( mainWith
+  , scienceOptions
+  , socialStudiesOptions
+  ) where
+```
+
+## Declaring Extensions
+
+- Declare extensions in the module(s) that use them, i.e. don't use
+  `default-extensions`
+
+  This ensures a) the source code works un-changed with other Haskell tooling
+  (e.g. `ghci`, `doctest`, etc) without further configuration and b) active
+  extensions are visible in the modules they impact.
+
+- Place extensions on their own line, and sort them
+
+  ```haskell
+  -- Bad
+  {-# LANGUAGE OverloadedStrings, RecordWildCards,
+      DataKinds #-}
+
+  -- Good
+  {-# LANGUAGE DataKinds #-}
+  {-# LANGUAGE OverloadedStrings #-}
+  {-# LANGUAGE RecordWildCards #-}
+  ```
+
+## Yesod
+
+When well-established Yesod idioms clash with the rules of this styleguide (e.g.
```

Comment:
When does a well-established yesod idiom clash with the style guide? I thought we allowed `default-extensions`?

Reply:
This patch adds the rule (with justification) to not use `default-extensions`, hence that being my clash example here.

The scaffold also generally does not follow our style, it does a lot of vertical alignment for example, and I wouldn't want to feel pressure to re-format the world (at least while we don't have a viable auto-formatter) after using the scaffold.

Reply:
Ah I see what you mean now. Honestly I wouldn't want to codify this because it creates some ambiguity. Also, we can change the scaffold however we want at this point, generating it was a one-time thing.

Reply:
Agreed, screw yesod defaults. We aren't using the scaffold tool anyway.